src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -239,10 +241,30 @@ private static Task<T> WithCancellation<T>(Task<T> task, CancellationToken cance
 
             if (cancellationToken.IsCancellationRequested)
             {
-                return new Task<T>(() => default(T), cancellationToken);
+                return Task.FromCanceled<T>(cancellationToken);
             }
 
-            return task.ContinueWith(t => t, cancellationToken, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default).Unwrap();
+            return withCancellationSlow(task, cancellationToken);
+
+            static async Task<T> withCancellationSlow(Task<T> task, CancellationToken cancellationToken)
+            {
+                Debug.Assert(task != null);
+                Debug.Assert(cancellationToken.CanBeCanceled);
+
+                var tcs = new TaskCompletionSource<bool>();
+                using (cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).TrySetResult(true), tcs))
+                {
+                    if (task != await Task.WhenAny(task, tcs.Task).ConfigureAwait(false))
```

Comment:
This code used to not `await`, but is flagged by an analyzer that tracks `await` and the solution is to add an `await.`. That doesn't seem right at all.